runtime/src/installed_scheduler_pool.rs

```diff
@@ -0,0 +1,446 @@
+//! special handling as part of their life-cycle.
+//!
+//! Albeit being at this abstract interface level, it's generally assumed that each
+//! `InstalledScheduler` is backed by multiple threads for performant transaction execution and
+//! there're multiple independent schedulers inside a single instance of `InstalledSchedulerPool`.
+//!
+//! Dynamic dispatch was inevitable, due to the need of delegating those implementations to the
+//! dependent crate (`solana-scheduler-pool`, which in turn depends on `solana-ledger`; another
+//! dependent crate of `solana-runtime`...), while cutting cyclic dependency.
+
+#[cfg(any(test, feature = "test-in-workspace"))]
+use mockall::automock;
+use {
+    crate::{bank::Bank, bank_forks::BankForks},
+    log::*,
+    solana_program_runtime::timings::ExecuteTimings,
+    solana_scheduler::{SchedulingMode, WithSchedulingMode},
+    solana_sdk::{
+        slot_history::Slot,
+        transaction::{Result, SanitizedTransaction},
+    },
+    std::{fmt::Debug, ops::Deref, sync::Arc},
+};
+
+// Send + Sync is needed to be a field of BankForks
+#[cfg_attr(any(test, feature = "test-in-workspace"), automock)]
+pub trait InstalledSchedulerPool: Send + Sync + Debug {
+    fn take_from_pool(&self, context: SchedulingContext) -> SchedulerBox;
+    fn return_to_pool(&self, scheduler: SchedulerBox);
+}
+
+#[cfg_attr(any(test, feature = "test-in-workspace"), automock)]
+// suppress false clippy complaints arising from mockall-derive:
+//   warning: `#[must_use]` has no effect when applied to a struct field
+#[allow(unused_attributes)]
+// Send + Sync is needed to be a field of Bank
+pub trait InstalledScheduler: Send + Sync + Debug {
+    fn scheduler_id(&self) -> SchedulerId;
+    fn scheduler_pool(&self) -> SchedulerPoolArc;
+
+    // Calling this is illegal as soon as schedule_termiantion is called on &self.
+    fn schedule_execution(&self, sanitized_tx: &SanitizedTransaction, index: usize);
+
+    // This optionally signals scheduling termination request to the scheduler.
+    // This is subtle but important, to break circular dependency of Arc<Bank> => Scheduler =>
+    // SchedulingContext => Arc<Bank> in the middle of the tear-down process, otherwise it would
+    // prevent Bank::drop()'s last resort scheduling termination attempt indefinitely
+    fn schedule_termination(&mut self);
+
+    #[must_use]
+    fn wait_for_termination(&mut self, reason: &WaitReason) -> Option<ResultWithTimings>;
+
+    // suppress false clippy complaints arising from mockall-derive:
+    //   warning: the following explicit lifetimes could be elided: 'a
+    #[allow(clippy::needless_lifetimes)]
+    fn scheduling_context<'a>(&'a self) -> Option<&'a SchedulingContext>;
+    fn replace_scheduler_context(&mut self, context: SchedulingContext);
+}
+
+pub type SchedulerPoolArc = Arc<dyn InstalledSchedulerPool>;
+pub(crate) type InstalledSchedulerPoolArc = Option<SchedulerPoolArc>;
```

Comment:
I might be missing some context.
It still feels strange to have the word `Installed` here, even when the same object is used while both being installed into a `Bank` and not.
In other words, an `InstalledSchedulerPoolArc`, on the first glance, would then need to refer to a pool of schedulers that are actually "installed".  But, IIUC, they are not.  It is holding schedulers that are not installed into any bank, and are waiting to be installed.

When I was trying to build a mental model of the abstractions interacting here, also reading your comments on the `Bank`/`Scheduler` circular references affecting their lifetimes, I wondered if choosing an ownership model could simplify both references and names.

Consider this:
1. Rename the `InstalledScheduler` trait to be just `Scheduler`.
2. Remove reference to the `Bank` from the `Scheduler`.
3. Only `Scheduler::execute` requires a `Bank`, so just pass it in as an argument.
4. When `Bank` will be using the scheduler it currently has installed, it will just pass `self` as the first argument to `execute()`.

I wonder if it would also remove the need for tracking the shutdown process, simplifying the `Scheduler` trait to this:

```rust
pub trait Scheduler: Send + Sync + Debug {
    fn id(&self) -> SchedulerId;
    fn pool(&self) -> InstalledSchedulerPoolArc;

    fn execute(&self, bank: &Bank, sanitized_tx: &SanitizedTransaction, index: usize);

    // suppress false clippy complaints arising from mockall-derive:
    //   warning: the following explicit lifetimes could be elided: 'a
    #[allow(clippy::needless_lifetimes)]
    fn context<'a>(&'a self) -> Option<&'a SchedulingContext>;

    fn replace_context(&mut self, context: SchedulingContext);
}
```

`Bank` now holds `Box<dyn Scheduler>` and returns it into a pool when it is frozen.

Reply:
> In other words, an `InstalledSchedulerPoolArc`, on the first glance, would then need to refer to a pool of schedulers that are actually "installed". But, IIUC, they are not. It is holding schedulers that are not installed into any bank, and are waiting to be installed.

how about `Installable`?



> When I was trying to build a mental model of the abstractions interacting here, also reading your comments on the `Bank`/`Scheduler` circular references affecting their lifetimes, I wondered if choosing an ownership model could simplify both references and names.

really, appreciate going so deep and suggestion. I think i mislead you. I think i can improve docs with this feedback from someone with non-perfect context

> Only Scheduler::execute requires a Bank, so just pass it in as an argument.
> ...
> ```rust
>     fn execute(&self, bank: &Bank, sanitized_tx: &SanitizedTransaction, index: usize);
> ````

admittedly, this isn't documented anywhere in this pr, but we need to make schedule_execution/execute non-blocking without executing it. soon, it just `crossbeam::Channel::send()` `SanitizedTransaction`. Also, `Scheduler`s are expected to be inherited across banks. so, we can't naitvely tie any SanitizedTransaction with particular bank at the time of scheduling. Also, if we were to tie, it will be forced to use `sync::Weak` to avoid bank leaks just because of buffered sanitized txes, which in turn incurs perf. penalty for each tx, which isn't acceptable.

all in all, `SchedulingContext` is born.

That said, i'm still open better trait design. :)

again, I'll write a proper in-source doc explaining this.

Reply:
> [... ] I think i mislead you. I think i can improve docs with this feedback from someone with non-perfect context

Glad my attempts to understand the PR brought at least some value :))
 
> > Only Scheduler::execute requires a Bank, so just pass it in as an argument.
> > ...
> > ```rust
> >     fn execute(&self, bank: &Bank, sanitized_tx: &SanitizedTransaction, index: usize);
> > ```
> 
> admittedly, this isn't documented anywhere in this pr, i thought this approach; but we can't take it.
> 
> we need to make `{schedule_execution/execute}()` non-blocking without executing txes. soon, it just `crossbeam::Channel::send()s` `SanitizedTransaction`s. Also, `Scheduler`s are expected to be inherited across banks in the case of block_generation scheduling_mode. so, we can't naitvely tie any SanitizedTransaction with particular bank at the time of scheduling. Also, if we were to tie, it will be forced to use `sync::Weak` to avoid bank leaks just because of buffered sanitized txes, which in turn incurs perf. penalty for each tx, which isn't acceptable.
> 
> all in all, `SchedulingContext` and complicated dance stage is born.
> 
> That said, i'm still open to better trait design. :)
> 
> again, I'll write a proper in-source doc explaining this.

I should have checked.  So the dependency is not via a `Bank` directly, but via a `SchedulingContext`.  I am still missing a complete picture, but just to be clear, you are saying that this:

> ```rust
>     fn execute(&self, context: &SchedulingContext, sanitized_tx: &SanitizedTransaction, index: usize);
> ```

Is not going to work, correct?
Is it because, the caller of the `execute()` does not have a `SchedulingContext` at hand, and you do not want to add it into the `Bank`?

Currently, `SchedulingContext` just wraps a reference to a `Bank`, adding a `mode` to it:

```rust
#[derive(Clone, Debug)]
pub struct SchedulingContext {
    mode: SchedulingMode,
    // Intentionally not using Weak<Bank> for performance reasons
    bank: Arc<Bank>,
}
```

Maybe if you put `mode` into the `Scheduler`, but keep the `Bank` as the `execute()` argument, it would work, ownership wise?
And if, in the future, the scheduling machinery is going to change, you can replace `mode` with a channel or anything else, still living inside a `Scheduler`, but the `Bank` is kept outside.
`Bank` and any other data is the only combined when `execute()` is invoked, and the reference to `Bank` is unnecessary for other method calls.
This way, there is no circular dependency between the `Bank` that (temporarily) owns the `Scheduler` and the `Scheduler` itself.

Something like this:

```rust
pub trait Scheduler: Send + Sync + Debug {
    fn id(&self) -> SchedulerId;
    fn pool(&self) -> InstalledSchedulerPoolArc;
    
    fn mode(&self) -> SchedulingMode;
    fn set_mode(&mut self, v: SchedulingMode);

    fn execute(&self, bank: &Bank, sanitized_tx: &SanitizedTransaction, index: usize);
}
```

Depending on how `mode` is used in a specific implementation, `mode()`/`set_mode()` might be unnecessary in this trait.  If, for example, setter and getter of the `mode` both know the concrete type of the `Scheduler` they work with.

Reply:
@ilya-bobyr hey, sorry for quite delayed reply. a good news is that i prepared benches, feeling concrete code communicates better.



> just to be clear, you are saying that this:
> 
> > ```rust
> >     fn execute(&self, context: &SchedulingContext, sanitized_tx: &SanitizedTransaction, index: usize);
> > ```
> 
> Is not going to work, correct?

no, it isn't. that's because `.execute()` must be nonblocking. worker threads must somehow be able to access to `&Bank` at later. (see the bench code)


> Maybe if you put `mode` into the `Scheduler`, but keep the `Bank` as the `execute()` argument, it would work, ownership wise?
> And if, in the future, the scheduling machinery is going to change, you can replace `mode` with a channel or anything else, still living inside a `Scheduler`
> ...
> Depending on how `mode` is used in a specific implementation, `mode()`/`set_mode()` might be unnecessary in this trait. If, for example, setter and getter of the `mode` both know the concrete type of the `Scheduler` they work with.

well, i want to tightly couple mode with any given bank. semantically, any given bank's scheduling mode must not be changed. so, that's why the tiny SchedulingContext is born.


> Something like this:
> 
> ```rust
> pub trait Scheduler: Send + Sync + Debug {
>     ...
> }
> ```

thanks for suggesting the simplified interface. i also wish things are that easy. could you take a look at those benches? 4c2df8dbbaaec6278ed7f88a74780ec2e25e775f (prep commit: 305e75b5f895de0d5ec4807a5d95b647bbbc1e5c). i hope the nonblocking requirement and resultant a bit complicated trait can be well justified then.

basically, any trait api simplication must beat the result of `bench_with_16_threads_without_arc_mutation`.

for reference, these are bench results:

```
     Running benches/scheduler.rs (target/x86_64-unknown-linux-gnu/release/deps/scheduler-48641774b6eafdf6)

running 10 tests
test blocking::bench_with_arc_mutation                       ... bench: 271,933,206 ns/iter (+/- 3,602,808)
test blocking::bench_without_arc_mutation                    ... bench: 258,657,392 ns/iter (+/- 2,889,026)
test nonblocking::bench_with_01_thread_with_arc_mutation     ... bench: 272,239,954 ns/iter (+/- 2,692,933)
test nonblocking::bench_with_01_thread_without_arc_mutation  ... bench: 258,352,224 ns/iter (+/- 3,472,439)
test nonblocking::bench_with_04_threads_with_arc_mutation    ... bench: 181,947,366 ns/iter (+/- 807,124)
test nonblocking::bench_with_04_threads_without_arc_mutation ... bench:  67,807,742 ns/iter (+/- 329,085)
test nonblocking::bench_with_08_threads_with_arc_mutation    ... bench: 156,013,390 ns/iter (+/- 323,350)
test nonblocking::bench_with_08_threads_without_arc_mutation ... bench:  35,974,832 ns/iter (+/- 192,139)
test nonblocking::bench_with_16_threads_with_arc_mutation    ... bench: 123,510,933 ns/iter (+/- 679,494)
test nonblocking::bench_with_16_threads_without_arc_mutation ... bench:  19,385,104 ns/iter (+/- 110,681)
```

(i needed the below hack to fully exercise the used powerful machine)

```
diff --git a/scheduler-pool/benches/scheduler.rs b/scheduler-pool/benches/scheduler.rs
index 1dfd46b227..918142f9b9 100644
--- a/scheduler-pool/benches/scheduler.rs
+++ b/scheduler-pool/benches/scheduler.rs
@@ -58,7 +58,7 @@ impl<const MUTATE_ARC: bool> ScheduledTransactionHandler<TransactionWithIndexFor
     ) {
         //std::hint::black_box(bank.clone());
         let mut i = 0;
-        for _ in 0..10 {
+        for _ in 0..100 {
             if MUTATE_ARC {
                 //for _ in 0..2 {
                 std::hint::black_box((Arc::downgrade(bank)).upgrade().unwrap());

```

Reply:
also, the above benches are also demonstrating why `std::sync::Weak` can't be used due to perf. concern, justifying yet another related inevitable introduction of cyclic ref of `Bank`.

Reply:
❗ arc downgrade + upgrade takes 1us? Did I do that math right? ~100_000_000 ns overhead per iter with 10_000 txs, 10 upgrade/downgrade per tx = 100_000 upgrade/downgrade per iter.

~4GHz machine => ~4000 cycles per upgrade downgrade?

That seems really high to me, maybe my understanding of how Arc works is not matched with reality.

Reply:
> exclamation arc downgrade + upgrade takes 1us?

well, you did that math wrong... it's actually 100ns (for the contended testcase):

> 100_000_000 ns overhead per iter with 10_000 txs, *100* (see my out-of-tree patch comment) upgrade/downgrade per tx = 1_000_000 upgrade/downgrade per iter.

that said, 100ns is still too costly, considering this perf. penalty increases as the count count goes up.

as shown by `bench_{with,without}_arc_mutation` and `bench_with_01_thread_{with,without}_arc_mutation`, the uncontended-case of overhead is ~14_000_000 ns, so 14ns. (that's reasonable for a few of cas) that's 6x slowdown, which can be completely avoidable with proper impl like this pr.

Reply:
@ilya-bobyr 

(hi, really thanks for another dense code-review round. I'll address them in order.)

So, the `InstalledScheduler` trait namaing and api is okay for you eyes by now?

I'm planning to start intense coding/writing session once it's okay as laid out at : https://github.com/solana-labs/solana/pull/31239#issuecomment-1562955659

Reply:
btw, I'm clinging to `AttachedSchduler` now that we're settling on `BankWithSchduler` while `InstalledSchedulerPool` is remaining as-is... lol